### PR TITLE
Done with UserRepositoryTask.

### PR DESCRIPTION
**Pull request by AlexanderVladimirov9090**
Done.

### CODE REVIEW

**Comment by freekman** [comment]
when you make a PR assign your mentor

File: source/com/clouway/userrepository/DataBaseUnreachableException.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
```
Comment (freekman) [inline source/com/clouway/userrepository/DataBaseUnreachableException.java]: Who told you to make the template like that 
it should be like 
```js
/**
 * @author ..... <...>
 */
````

File: UserRepositoryScenario [hunk 2]
```diff
@@ -0,0 +1,15 @@
+UserRepository.
```
Comment (freekman) [inline UserRepositoryScenario]: what is the purpose of this file

File: source/com/clouway/userrepository/User.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
+ *         (Alexander Vladimirov)
+ */
+public class User {
+  private final String age;
```
Comment (freekman) [inline source/com/clouway/userrepository/User.java]: if you make the field ```public final```` you could remove the gethers

File: source/com/clouway/userrepository/UserDataBase.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
+ *         (Alexander Vladimirov)
+ */
+public interface UserDataBase {
```
Comment (freekman) [inline source/com/clouway/userrepository/UserDataBase.java]: for the interfaces it is considered good practice to add documentation

File: source/com/clouway/userrepository/UserDataBase.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+    boolean add(User user)throws DataBaseUnreachableException;
+    boolean userExists(String name)throws UserExistsException,UserDoesNotExistException;
```
Comment (freekman) [inline source/com/clouway/userrepository/UserDataBase.java]: please format

File: source/com/clouway/userrepository/UserDoesNotExistException.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
+ *         (Alexander Vladimirov)
+ */
+public class UserDoesNotExistException extends Exception {
```
Comment (freekman) [inline source/com/clouway/userrepository/UserDoesNotExistException.java]: you have ```UserDoesNotExistException``` and ```UserExistsException``` is there a difference between them

File: source/com/clouway/userrepository/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
+ *         (Alexander Vladimirov)
+ */
+class UserRepository {
+  private final UserDataBase dataBase;
+  private final UserValidator validator;
+
+
```
Comment (freekman) [inline source/com/clouway/userrepository/UserRepository.java]: .

File: source/com/clouway/userrepository/UserValidator.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author alexandervladimirov1902@gmail.com
+ *         (Alexander Vladimirov)
+ */
+public interface UserValidator {
```
Comment (freekman) [inline source/com/clouway/userrepository/UserValidator.java]: .

File: source/com/clouway/userrepository/UserRepository.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+          return true;
+        }
+        throw new UserExistsException();
+      }
+    }
+    return false;
+  }
+
+  public boolean isUserAdult(String name) throws UserDoesNotExistException, UserExistsException {
+    if (!dataBase.userExists(name)) {
```
Comment (mgenov) [inline source/com/clouway/userrepository/UserRepository.java]: There is a smell in the design of this code, cause you will make 2 database calls for single thing. 
1. Checking whether user exists 
2. Retrieving user  

Actually both a partially overlapped. 

```java
User user = database.findUserByName(name); 
if (user == null) {
   throw new UserDoesNotExistsException();
}
if (user.age() < 18) {
  return false;
}
return true;
```

or better with optional

```java
Optional<User> possibleUser = database.findUserByName(name); 
if (!possibleUser.isPresent()) {
   throw new UserDoesNotExistsException();
}
User user = possibleUser.get();
if (user.age() < 18) {
  return false;
}
return true;
```

Comment (mgenov) [inline source/com/clouway/userrepository/UserRepository.java]: Null explained in details: 
https://github.com/google/guava/wiki/UsingAndAvoidingNullExplained

Optional is part of java 8. 
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

File: source/com/clouway/userrepository/DataBaseUnreachableException.java (hunk 1 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/DataBaseUnreachableException.java]: Fixed it, next task will be with the right template.

File: UserRepositoryScenario (hunk 2 above)
Comment (AlexanderVladimirov9090) [inline UserRepositoryScenario]: This is my scenario file I`ll remove it.

File: source/com/clouway/userrepository/User.java (hunk 3 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/User.java]: Will be done.

File: source/com/clouway/userrepository/UserDataBase.java (hunk 4 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserDataBase.java]: Will be done.

File: source/com/clouway/userrepository/UserDataBase.java (hunk 5 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserDataBase.java]: Yeah I see what you are saying.

File: source/com/clouway/userrepository/UserDoesNotExistException.java (hunk 6 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserDoesNotExistException.java]: Will be fixed. 
I was going for when someone is trying to register user when exists UserExistsException will be thrown. 
When someone is trying to get some user by isAdult() that is not in Data Base the same method will throw UserDoesNotExistsException.

File: source/com/clouway/userrepository/UserRepository.java (hunk 9 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserRepository.java]: Today was reading about it, think it is clever.

**Comment by AlexanderVladimirov9090** [comment]
Done.

File: source/com/clouway/userrepository/DataBaseUnreachableException.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author Alexander Vladimirov
+ *        (alexandervladimirov1902@gmail.com)
+ */
+public class DataBaseUnreachableException extends Exception {
```
Comment (EmilGeorgiev) [inline source/com/clouway/userrepository/DataBaseUnreachableException.java]: Extend ```Exception```  sometimes is not well. Please read https://dzone.com/articles/how-to-properly-use-exceptions-in-java

File: source/com/clouway/userrepository/UserDataBase.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+   */
+  boolean exists(String name);
+
+  /**
+   * Retrieves user from Data Base.
+   *
+   * @param name using name to search user.
+   * @return user from Data Base.
+   */
+  User retrieve(String name);
```
Comment (EmilGeorgiev) [inline source/com/clouway/userrepository/UserDataBase.java]: retrieve(string) is not readable. Rename method to findByName(String)

File: source/com/clouway/userrepository/UserDoesNotExistException.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author Alexander Vladimirov
+ *        (alexandervladimirov1902@gmail.com)
+ */
+public class UserDoesNotExistException extends Exception {
```
Comment (EmilGeorgiev) [inline source/com/clouway/userrepository/UserDoesNotExistException.java]: You can common exception UserException for  ```UserDoesNotExistException``` and ```UserExistsException```.:

```java
public class UserException extends RuntimeException {
     public  UserException(String message) {
         super(mesage)
     }
}
```

File: source/com/clouway/userrepository/UserValidator.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package com.clouway.userrepository;
+
+/**
+ * Created by clouway on 06.07.16.
+ *
+ * @author Alexander Vladimirov
+ *        (alexandervladimirov1902@gmail.com)
+ * This user validator interface is for basic validation of user before added to Data Base.
+ */
+public interface UserValidator {
```
Comment (EmilGeorgiev) [inline source/com/clouway/userrepository/UserValidator.java]: make common method validate(User) which will validate age and name internally. Than this code ``` (validator.checkValidAge(user)&&validator.checkValidName(user))``` will be  ```(validator.Validate(sser))```

File: source/com/clouway/userrepository/UserRepository.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * Validates if user is adult.
+   *
+   * @param name is used to validate if is adult or not.
+   * @return true if user is adult and false if not.
+   * @throws UserDoesNotExistException when given name is not user in Data Base.
+   */
+  public boolean isUserAdult(String name) throws UserDoesNotExistException {
+    Optional<User> possibleUser = Optional.of(dataBase.retrieve(name));
```
Comment (EmilGeorgiev) [inline source/com/clouway/userrepository/UserRepository.java]: rename ```possibleUser``` to ```user```

File: source/com/clouway/userrepository/UserRepository.java (hunk 9 above)
Comment (mgenov) [inline source/com/clouway/userrepository/UserRepository.java]: cc @EmilGeorgiev

File: source/com/clouway/userrepository/UserRepository.java (hunk 14 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserRepository.java]: when I user retrieve and if the user is not there this possibleUser is not any kind of user for my perspective
That is why I think that possibleUser sounds good when mgenov proposed in the snippet above .

File: source/com/clouway/userrepository/UserValidator.java (hunk 13 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserValidator.java]: Will be done.

File: source/com/clouway/userrepository/UserDataBase.java (hunk 11 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/UserDataBase.java]: Point taken.

File: source/com/clouway/userrepository/DataBaseUnreachableException.java (hunk 10 above)
Comment (AlexanderVladimirov9090) [inline source/com/clouway/userrepository/DataBaseUnreachableException.java]: Understood.

**Comment by EmilGeorgiev** [comment]
LGTM

**Pull request merged**